packages/db/src/models/granule.ts

```diff
@@ -5,30 +5,79 @@ import { tableNames } from '../tables';
 
 import { PostgresGranule, PostgresGranuleRecord } from '../types/granule';
 
+import { GranuleExecutionHistoryPgModel } from './granule-execution-history';
+
 export default class GranulePgModel extends BasePgModel<PostgresGranule, PostgresGranuleRecord> {
   constructor() {
     super({
       tableName: tableNames.granules,
     });
   }
 
-  upsert(
+  async createWithExecutionHistory(
+    knexOrTrx: Knex | Knex.Transaction,
+    item: PostgresGranule,
+    executionCumulusId: number,
+    granuleExecutionHistoryPgModel = new GranuleExecutionHistoryPgModel()
+  ) {
+    const [granuleCumulusId] = await this.create(knexOrTrx, item);
+    await granuleExecutionHistoryPgModel.create(
+      knexOrTrx,
+      {
+        granule_cumulus_id: granuleCumulusId,
+        execution_cumulus_id: executionCumulusId,
+      }
+    );
+    return [granuleCumulusId];
+  }
+
+  getWithExecutionHistory(
+    knexOrTransaction: Knex | Knex.Transaction,
+    granule: Partial<PostgresGranuleRecord>
+  ) {
+    // TODO: is this open to SQL injection?
+    return knexOrTransaction(this.tableName)
+      .where(granule)
+      .join(
+        tableNames.granuleExecutionsHistory,
+        `${this.tableName}.cumulus_id`,
+        '=',
+        `${tableNames.granuleExecutionsHistory}.granule_cumulus_id`
+      )
+      .column(`${this.tableName}.*`)
+      .column(
+        knexOrTransaction.raw(`array_agg(${tableNames.granuleExecutionsHistory}.execution_cumulus_id) as execution_cumulus_ids`)
+      )
+      .groupBy('cumulus_id');
+  }
+
+  async upsert(
     knexOrTrx: Knex | Knex.Transaction,
-    granule: PostgresGranule
+    granule: PostgresGranule,
+    executionCumulusId: number,
+    granuleExecutionHistoryPgModel = new GranuleExecutionHistoryPgModel()
   ) {
     if (granule.status === 'running') {
       return knexOrTrx(this.tableName)
         .insert(granule)
         .onConflict(['granule_id', 'collection_cumulus_id'])
         .merge({
-          execution_cumulus_id: granule.execution_cumulus_id,
           status: granule.status,
           timestamp: granule.timestamp,
           updated_at: granule.updated_at,
         })
-        // execution_cumulus_id is not required, so granule.execution_cumulus_id may be
-        // undefined. so need to compare against EXCLUDED.execution_cumulus_id
-        .whereRaw(`${this.tableName}.execution_cumulus_id != EXCLUDED.execution_cumulus_id`)
+        // Only do the upsert if there IS NOT already a record associating
+        // the granule to this execution. If there IS already a record
+        // linking this granule to this execution, then this upsert query
+        // will not affect any rows.
+        .whereNotExists(
+          // TODO: this effectively makes executions required for granules
```

Comment:
Since the execution cumulus ID is used to look up whether there are any existing records associating this granule to this execution, it becomes required. I might be able to refactor this query to make it non-required, but the more I think about it, it kind of feels to me like it should be required. Is there any scenario where we don't want to have an execution to associate with a granule? 

Also, for the PG writes, if the granule write has even been reached, there should have been a successful execution write to PG.

This is a breaking change relative to the previous implementation.

Reply:
That said.... are we saying we don't want to update the granule fields if there *isn't* a granule-execution link found?   If so, this will return `[]` if there is a link missing, and that almost feels like it should be constrained by an error.   

I would think this logic should be an analogue to create, in that if the join record is missing we create it - and the reason being, for example,  is it's *possible* for workflows with a database update step mid workflow to apply RUNNING states out of order on the same execution and metadata might be updated/matter in that instance.

Reply:
Let's discuss this offline